src/Strategy/Haskell/Stack.hs

```diff
@@ -0,0 +1,127 @@
+newtype PackageName = PackageName {unPackageName :: Text} deriving (FromJSON, Eq, Ord, Show)
+newtype StackLocation = StackLocation {unStackLocation :: StackLocationType} deriving (Eq, Ord, Show)
+
+data StackDep = StackDep
+  { stackName :: PackageName,
+    stackVersion :: Text,
+    stackDepNames :: [PackageName],
+    stackLocation :: StackLocationType
+  }
+  deriving (Eq, Ord, Show)
+
+data StackLocationType
+  = Local
+  | Remote
+  | BuiltIn
+  deriving (Eq, Ord, Show)
+
+instance FromJSON StackDep where
+  parseJSON = withObject "StackDep" $ \obj ->
+    StackDep <$> obj .: "name"
+      <*> obj .: "version"
+      <*> obj .:? "dependencies" .!= []
+      <*> fmap unStackLocation (obj .:? "location" .!= StackLocation BuiltIn)
+
+instance FromJSON StackLocation where
+  parseJSON = withObject "StackLocation" $ \obj ->
+    StackLocation <$> (obj .: "type" >>= parseLocationType)
+
+parseLocationType :: MonadFail m => Text -> m StackLocationType
+parseLocationType txt
+  | txt == "hackage" = pure Remote
+  | txt `elem` ["project package", "archive"] = pure Local
+  | otherwise = fail $ "Bad location type: " ++ T.unpack txt
+
+discover :: HasDiscover sig m => Path Abs Dir -> m ()
+discover = walk $ \dir _ files ->
+  case find (\f -> fileName f == "stack.yaml") files of
+    Nothing -> pure WalkContinue
+    Just _ -> do
+      runSimpleStrategy "haskell-stack" HaskellGroup $ fmap (mkProjectClosure dir) (analyze dir)
+      pure WalkSkipAll
+
+stackJSONDepsCmd :: Command
+stackJSONDepsCmd =
+  Command
+    { cmdName = "stack",
+      cmdArgs = ["ls", "dependencies", "json"],
+      cmdAllowErr = Never
+    }
+
+mkProjectClosure :: Path Abs Dir -> G.Graphing Dependency -> ProjectClosureBody
+mkProjectClosure dir graph =
+  ProjectClosureBody
+    { bodyModuleDir = dir,
+      bodyDependencies = dependencies,
+      bodyLicenses = []
+    }
+  where
+    dependencies =
+      ProjectDependencies
+        { dependenciesGraph = graph,
+          dependenciesOptimal = Optimal,
+          dependenciesComplete = Complete
+        }
+
+doGraph :: Has (MappedGrapher PackageName StackDep) sig m => StackDep -> m ()
+doGraph dep = do
+  let name = stackName dep
+  mapping name dep
+  for_ (stackDepNames dep) $ \child -> do
+    edge name child
+    when (stackLocation dep == Local) (direct child)
+
+ignorePackageName :: PackageName -> a -> a
+ignorePackageName = flip const
+
+shouldInclude :: StackDep -> Bool
+shouldInclude dep = Remote == stackLocation dep
+
+toDependency :: StackDep -> Dependency
+toDependency dep =
+  Dependency
+    { dependencyType = HackageType,
+      dependencyName = unPackageName $ stackName dep,
+      dependencyVersion = Just $ CEq $ stackVersion dep,
+      dependencyLocations = [],
+      dependencyEnvironments = [],
+      dependencyTags = M.empty
+    }
+
+buildGraph :: Has Diagnostics sig m => [StackDep] -> m (G.Graphing Dependency)
+buildGraph deps = do
+  result <- withMapping ignorePackageName $ traverse doGraph deps
```

Comment:
I just remembered we have `fromEither :: (ToDiagnostic err, Has Diagnostics sig m) => Either err a -> m a` in the Diagnostics module:

```hs
result <- fromEither =<< withMapping ignorePackageName (traverse doGraph deps)
pure . G.gmap toDependency . G.filter shouldInclude $ result
```

unsure if this is too clever